File: DSA-Python/3-maths/6-divisors.py
# Given an integer n, return all divisors of n


# brute force - O(n), O(1)
def divisors_1(n:int) -> list:
    divs = []
    
    for i in range(1, n+1):
        if (n % i == 0):
            divs.append(i)
            
    return divs


# optimal - O(rt(n)), O(1)
def divisors_2(n:int) -> list:
    divs = []
    
    i=1
    while i*i <= n:
        if n % i == 0:
            divs.append(i)
            
            if i != n//i:
                divs.append(n//i)
                
        i+= 1
    
    return divs
    
    
if __name__=="__main__":
    n = int(input())
    
    for _ in range(n):
        n = int(input())
        
        res = divisors_1(n)
        # res = divisors_2(n)
        
        print(sorted(res))
        
'''
5
36
12
0
13
14
'''

File: DSA-C++/2-maths/6-divisors.cpp
#include <bits/stdc++.h>

using namespace std;

// brute force - O(n), O(n)
vector <int> divisors_1(int num) {
    vector <int> ans;

    for (int i=1; i<=num; ++i) {
        if (num % i == 0) {
            ans.push_back(i);
        }
    }

    return ans;
}

// optimal - O(sqrt(n)), O(n)
vector <int> divisors_2(int num) {
    vector <int> ans;

    for (int i=1; i*i<=num; ++i) {
        if (num % i == 0) {
            ans.push_back(i);

            if (i != num/i) {
                ans.push_back(num/i);
            }
        }
    }

    return ans;
}


int main() {
    int num = 36;

    for (int x:divisors_1(num)) cout << x << " ";
    
    // for (int x:divisors_2(num)) cout << x << " ";

    cout << endl;
    return 0;
}